Scripts/Runtime/Jobs/Deferred/IJobDeferredNativeArrayFor.cs

```diff
@@ -0,0 +1,142 @@
+using Anvil.Unity.DOTS.Collections;
+using System;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using Unity.Burst;
+using Unity.Collections;
+using Unity.Collections.LowLevel.Unsafe;
+using Unity.Jobs;
+using Unity.Jobs.LowLevel.Unsafe;
+using UnityEngine.Scripting;
+
+namespace Anvil.Unity.DOTS.Jobs
+{
+    /// <summary>
+    /// A replacement for <see cref="IJobFor"/> when the number of work items is not known at Schedule time
+    /// and you are using a <see cref="DeferredNativeArray{T}"/>
+    /// </summary>
+    [JobProducerType(typeof(JobDeferredNativeArrayForExtensions.JobDeferredNativeArrayForProducer<>))]
+    public interface IJobDeferredNativeArrayFor
+    {
+        /// <summary>
+        /// Implement this method to perform work against a specific iteration index.
+        /// </summary>
+        /// <param name="index">The index of the <see cref="NativeArray{T}"/> from a <see cref="DeferredNativeArray{T}"/></param>
+        void Execute(int index);
+    }
+
+    public static class JobDeferredNativeArrayForExtensions
+    {
+        public static unsafe JobHandle ScheduleParallel<TJob, T>(this TJob jobData,
+                                                                 DeferredNativeArray<T> deferredNativeArray,
+                                                                 int batchSize,
+                                                                 JobHandle dependsOn = default)
+            where TJob : struct, IJobDeferredNativeArrayFor
+            where T : struct
+        {
+            void* atomicSafetyHandlePtr = null;
+
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
+            atomicSafetyHandlePtr = DeferredNativeArrayUnsafeUtility.GetSafetyHandlePointer(ref deferredNativeArray);
+#endif
+
+            IntPtr reflectionData = JobDeferredNativeArrayForProducer<TJob>.s_JobReflectionData.Data;
+            CheckReflectionDataCorrect(reflectionData);
```

Comment:
🚲 Super minor but maybe this check be in a getter on `JobDeferredNativeArrayForProducer<TJob>`?

Reply:
It's a static method so it can be removed with the `[Conditional("ENABLE_UNITY_COLLECTIONS_CHECKS")]`

I was just following suit with how Unity does it for their job scheduling.

Reply:
I meant making the call to `CheckReflectionDataCorrect` in the getter instead of after getting so all who read the value have to pass the check. Would still get discarded on the conditional.

If you'd rather follow Unity that's ok.

Reply:
Fixed.